### ISSUE TITLE
2FA Login: Offer the option to enter a backup code

### ISSUE THREAD

**Issue by thehenrybyrd** [wordpress-mobile/WordPress-Android#15961]
### Expected behavior
When logging into an account with 2FA enabled (whether via an Auth app or SMS), we should clearly show the option to enter a backup code.

### Actual behavior
Currently, whether the user's account has Auth app 2FA or SMS 2FA, we ask them to "enter the verification code from your Authenticator app", and give the option to "Text me a code instead".

This implies that all users are using an Auth app, and that SMS 2FA is an alternative. In reality, recent WordPress.com accounts use either one or the other, not both.
Part of making this clearer would be to display different messages based on which kind of 2FA the user's account uses (https://github.com/wordpress-mobile/WordPress-Android/issues/11746).

However, as @guarani pointed out in beta testing for 19.2 (p5T066-2Yv#comment-11230), if the user's primary 2FA method isn't working for them, they'll need to enter a backup code.

### Steps to reproduce the behavior
1. Set up 2FA on WordPress.com, using either an Auth app or SMS. 
2. Login to the WordPress app, but pretend you don't have access to your Auth app or can't receive the SMS.
3. Note that we do not mention the possibility to enter a backup code.

**Comment by guarani** [comment]
Thanks for creating this issue!

> However, as @guarani pointed out in beta testing for 19.2 (p5T066-2Yv#comment-11230), if the user's primary 2FA method isn't working for them, they'll need to enter a backup code.

I hadn't realized that it was actually possible to enter a backup code instead of a 2FA code. I tested now and it does work.

---

I wonder if the simplest way of fixing this would be to change:

> Please enter the verification code from your Authenticator app.

to:

> Please enter the verification code from your Authenticator app or enter a backup code.

Alternatively, we could add another text link below "Text me a code instead" that says "Try another way". Tapping it could show an alert with a message like "Please enter a back code."

**Comment by thehenrybyrd** [comment]
@guarani I think changing the wording is a great way to approach this issue! 
We could probably reduce the confusion associated with https://github.com/wordpress-mobile/WordPress-Android/issues/11746 in the same manner.

For instance we could change

> Please enter the verification code from your Authenticator app.

to something like

> Please enter your verification code or a backup code.

(though we'd need to be careful with the wording - it might be helpful to users to read "from your Authenticator app" or "that we've sent via text message")

We could also change
> "Text me a code instead"

to 

> "Text me another code"

to account for the situation where someone with 2FA via SMS doesn't get the code initially.

However, if we are going to change that we should also check for edge cases like old accounts that might have both Auth and SMS (I feel like this used to be possible, but I'm not certain).

**Comment by guarani** [comment]
@thehenrybyrd, the above suggestions sound great! https://github.com/wordpress-mobile/WordPress-Android/issues/11746 and this are closely related, so it makes sense to think of them together here.

> For instance we could change
> 
> > Please enter the verification code from your Authenticator app.
> 
> to something like
> 
> > Please enter your verification code or a backup code.
> 
> (though we'd need to be careful with the wording - it might be helpful to users to read "from your Authenticator app" or "that we've sent via text message")

I personally find value in the messaging being specific e.g. "from your Authenticator app" or "that we've sent via text message", but we could go with the more generic messaging "Please enter your verification code or a backup code." you mention. We could probably also link to this discussion on https://github.com/wordpress-mobile/WordPress-Android/issues/11746 for visibility.

---

> We could also change
> 
> > "Text me a code instead"
> 
> to
> 
> > "Text me another code"

I'm not sure we currently have the means to re-send an SMS code. When logging in with a user account that has SMS 2FA enabled, the current "Text me a code instead" button doesn't seem to work. I don't receive another SMS and the Activity Log shows "Failed to request one time code". There might well be way to re-send an SMS from the app, but it might require more investigation. 

Re-requesting SMS codes feels like it could be a separate issue (a separate feature request), since it seems independent of this issue (messaging around backup codes) and https://github.com/wordpress-mobile/WordPress-Android/issues/11746 (messaging around specific 2FA methods), would you agree? I'm happy to create an issue if you think this is a good idea.

**Comment by thehenrybyrd** [comment]
@guarani I see what you mean about more specific messaging:

> I personally find value in the messaging being specific e.g. "from your Authenticator app" or "that we've sent via text message", but we could go with the more generic messaging

If we do end up updating the API endpoint as discussed in https://github.com/wordpress-mobile/WordPress-Android/issues/11746, we could likely address the messaging with that or a related change. If we decide the backend work is too much, then more generic wording here could be an easier solution. 

As for re-requesting SMS, I agree that it's a separate feature request. It's a nice idea, however I'm not sure how frequently users need to re-request an SMS.

**Comment by guarani** [comment]
> If we do end up updating the API endpoint as discussed in #11746, we could likely address the messaging with that or a related change. If we decide the backend work is too much, then more generic wording here could be an easier solution.

Makes perfect sense 👍 

> As for re-requesting SMS, I agree that it's a separate feature request. It's a nice idea, however I'm not sure how frequently users need to re-request an SMS.

I'm not sure either if it's an issue users run into often. I have ran into an issue myself where the SMS doesn't arrive and it appears to be because I had one sent to me within the last minute. I'd changed the account password on WordPress.com (which required an SMS 2FA code) and then logged into WPAndroid straight afterwards. The code arrived for the WP.com change password flow, but didn't for the WPAndroid login. Having a button to re-send the code might have helped here. That said, in my opinion having the app display a message if the SMS limit is exceeded would be higher priority than having the re-send option.

| Message displayed on WP.com when SMS limit is exceeded |
| - |
| <img src="https://user-images.githubusercontent.com/1898325/154955626-5d34d705-1a6f-44a2-8baa-9aeff8ccc810.png" alt="" width="350">  |

In summary, would you agree we should create two more issues for WPiOS and WPAndroid? They are:
- Enhancement request to alert users if SMS limit is reached
- Enhancement request to add a "re-send SMS" button on the SMS verification code screen

**Comment by thehenrybyrd** [comment]
> In summary, would you agree we should create two more issues for WPiOS and WPAndroid?

@guarani yes - that makes sense to me, can you create those? 
Thanks so much for all your thoughtful discussion here!

**Comment by guarani** [comment]
Thank you @thehenrybyrd for the discussion! It's been really helpful ❤️ 

> yes - that makes sense to me, can you create those?

Sure, I created issues for "Enhancement request to alert users if SMS limit is reached":
- https://github.com/wordpress-mobile/WordPress-iOS/issues/18069
- https://github.com/wordpress-mobile/WordPress-Android/issues/16035

I'll loop back here with issues for _'Enhancement request to add a "re-send SMS" button on the SMS verification code screen'_ 👍

**Comment by thehenrybyrd** [comment]
Thanks so much @guarani!

**Comment by guarani** [comment]
> I'll loop back here with issues for _'Enhancement request to add a "re-send SMS" button on the SMS verification code screen'_ 👍

To close the loop here: I came to realize that https://github.com/wordpress-mobile/WordPress-Android/issues/11746 covers this already. 
- As @thehenrybyrd [pointed out above](https://github.com/wordpress-mobile/WordPress-Android/issues/15961#issuecomment-1042741181), the  "Text me a code instead" _does_ resend the SMS on accounts with SMS-based 2FA. This didn't appear to work for me at first, but I was likely just running into the SMS rate limiting issue: https://github.com/wordpress-mobile/WordPress-Android/issues/16035.
- I added a comment to make sure this button copy change is part of the existing issue that covers the 2FA screen: https://github.com/wordpress-mobile/WordPress-Android/issues/11746#issuecomment-1111404732.

Please let me know if any of this seems inaccurate.

**Comment by thehenrybyrd** [comment]
@guarani that makes sense to me! I'll go ahead and close this issue in favor of #11746.

**Comment by guarani** [comment]
Thank you @thehenrybyrd!
